begin work on some homotopy coherences (#113)
rzk-lang/sHoTT: src/hott/02-homotopies.rzk.md

Homotopy composition is defined in diagrammatic order like `#!rzk concat` but
unlike composition.

The following is the unit for compositions of homotopies.

```rzk
#def refl-htpy
  (f : A → B)
  : homotopy f f
  := \ x → refl
```

```rzk
#end homotopies
```

There is also a dependent version of homotopy.

```rzk
#def dependent-homotopy
  ( A : U)
  ( B : A → U)
  (f g : (a : A) → B a)
  : U

>   := (a : A) → (f a = g a)

Reviewer: What you call a dependent homotopy here I would call just a homotopy. The notion of dependent homotopy should generalize dependent identifications. You can find a formalization of it here:

https://github.com/UniMath/agda-unimath/blob/591dd732c2bf6876871b941e952f88364c43ffe1/src/foundation/dependent-homotopies.lagda.md

Author: Indeed! But in rzk currently homotopies are not defined for dependent functions.

Author: @fredrik-bakke  Should I be the one to apply these changes globally? I'm happy to, but wanted to make sure.

Reviewer: That would be great, if you're comfortable with that

Author: Yup! Absolutely.

```

## Some path algebra for homotopies

> This section contains some lemmas on the groupoidal structure of homotopies.

Reviewer: ```suggestion
In this section we prove some lemmas on the groupoidal structure of homotopies.
```

Given a homotopy between two homotopies $C : H \sim K$ this produces a homotopy
$K^{-1} \cdot H  \sim \text{refl-htpy}_{g}$

```rzk
#def htpy-cancel-left
 (A B : U)
 (f g : A → B)
 (H K : homotopy A B f g)
 (C : dependent-homotopy A (\ a → f a = g a) H K)

>  : dependent-homotopy
>     A
>     (\ b → g b = g b)
>     (\ x → concat B (g x) (f x) (g x) (rev B (f x) (g x) (K x)) (H x))
>     (refl-htpy A B g)

Reviewer: ```suggestion
 : dependent-homotopy
    ( A)
    ( \ b → g b = g b)
    ( \ x → concat B (g x) (f x) (g x) (rev B (f x) (g x) (K x)) (H x))
    ( refl-htpy A B g)
```

Author: Thanks for the formatting suggestions. And indeed there is quite a bit left to work on this PR. I’ve marked it as a draft.

>   := \ x →
>       cancel-left-path
>       (B)
>       (f x)
>       (g x)
>       (H x)
>       (K x)
>       (C x)

Reviewer: ```suggestion
  := \ x → cancel-left-path B (f x) (g x) (H x) (K x) (C x)
```
When the code fits within the 80-character limit, it is usually preferable to write it on a single line.

      ( p)
```

It is sometimes useful to have this in inverse form.

```rzk
#def rev-nat-htpy
  ( A B : U)
  ( f g : A → B)
  ( H : homotopy A B f g)
  ( x y : A)
  ( p : x = y)
  : ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p)) =
    ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))

>   := rev
>       ( f x = g y)
>       ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))
>       ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p))
>       ( nat-htpy A B f g H x y p)

Reviewer: Always break line after walrus separator when definition is multiline
```suggestion
  :=
  rev
  ( f x = g y)
  ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))
  ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p))
  ( nat-htpy A B f g H x y p)
```
